File: dbgen/p07_healthcare.py
import duckdb, numpy as np, sys, os
from datetime import date, timedelta
from concurrent.futures import ProcessPoolExecutor
from utils.synth_utils import (
    GenerationProgress,
    batched_insert,
    get_worker_count,
    print_generation_summary,
    run_parallel,
)


def generate_patients_chunk(start, end, genders, plans, states, base):
    size = end - start
    rng = np.random.default_rng(start)
    days_back = rng.integers(365 * 5, 365 * 85 + 1, size)
    gender_indices = rng.integers(0, len(genders), size)
    zip_codes = rng.integers(10000, 100000, size)
    plan_indices = rng.integers(0, len(plans), size)
    state_indices = rng.integers(0, len(states), size)

    rows = []
    for idx, i in enumerate(range(start, end)):
        rows.append((
            i,
            base - timedelta(days=int(days_back[idx])),
            genders[gender_indices[idx]],
            str(zip_codes[idx]),
            plans[plan_indices[idx]],
            states[state_indices[idx]],
        ))
    return rows

def generate_providers_chunk(start, end, specialties, states):
    size = end - start
    rng = np.random.default_rng(start)
    spec_indices = rng.integers(0, len(specialties), size)
    state_indices = rng.integers(0, len(states), size)
    network_probs = rng.random(size)

    rows = []
    for idx, i in enumerate(range(start, end)):
        rows.append((
            i,
            f"Provider {i}",
            specialties[spec_indices[idx]],
            states[state_indices[idx]],
            bool(network_probs[idx] > 0.2),
            f"NPI{i:010d}",
        ))
    return rows

def generate_claims_chunk(start, end, NPA, NPR, base, ctypes, cstatuses, denial_reasons):
    size = end - start
    rng = np.random.default_rng(start)
    patient_ids = rng.integers(1, NPA + 1, size)
    provider_ids = rng.integers(1, NPR + 1, size)
    days_offset = rng.integers(0, 1096, size)
    ctype_indices = rng.integers(0, len(ctypes), size)
    bill_amounts = rng.uniform(100, 50000, size)
    allow_mults = rng.uniform(0.4, 0.95, size)
    paid_mults = rng.uniform(0.5, 1.0, size)
    paid_probs = rng.random(size)
    status_indices = rng.integers(0, len(cstatuses), size)
    denial_indices = rng.integers(0, len(denial_reasons), size)

    rows = []
    for idx, i in enumerate(range(start, end)):
        bill = round(float(bill_amounts[idx]), 2)
        allow = round(bill * float(allow_mults[idx]), 2)
        paid = round(allow * float(paid_mults[idx]) if paid_probs[idx] > 0.1 else 0.0, 2)
        rows.append((
            i,
            int(patient_ids[idx]),
            int(provider_ids[idx]),
            base + timedelta(days=int(days_offset[idx])),
            ctypes[ctype_indices[idx]],
            bill,
            allow,
            paid,
            cstatuses[status_indices[idx]],
            denial_reasons[denial_indices[idx]],
        ))
    return rows

def generate_claim_lines_chunk(start, end, NCL, cpt_codes):
    size = end - start
    rng = np.random.default_rng(start)
    claim_ids = rng.integers(1, NCL + 1, size)
    cpt_indices = rng.integers(0, len(cpt_codes), size)
    quantities = rng.integers(1, 6, size)
    unit_costs = rng.uniform(10, 5000, size)
    allowed_amounts = rng.uniform(5, 4000, size)
    paid_amounts = rng.uniform(0, 3500, size)

    rows = []
    for idx, i in enumerate(range(start, end)):
        rows.append((
            i,
            int(claim_ids[idx]),
            cpt_codes[cpt_indices[idx]],
            int(quantities[idx]),
            round(float(unit_costs[idx]), 2),
            round(float(allowed_amounts[idx]), 2),
            round(float(paid_amounts[idx]), 2),
        ))
    return rows

def generate_diagnoses_chunk(start, end, NCL, icd_codes):
    size = end - start
    rng = np.random.default_rng(start)
    claim_ids = rng.integers(1, NCL + 1, size)
    icd_indices = rng.integers(0, len(icd_codes), size)
    primary_probs = rng.random(size)
    chronic_probs = rng.random(size)

    rows = []
    for idx, i in enumerate(range(start, end)):
        rows.append((
            i,
            int(claim_ids[idx]),
            icd_codes[icd_indices[idx]],
            bool(primary_probs[idx] > 0.3),
            bool(chronic_probs[idx] > 0.6),
        ))
    return rows


def main():
    sf = float(sys.argv[1]) if len(sys.argv) > 1 else 1.0
    sf_adj = sf * 1000.0
    NPA = max(20, int(1000 * sf_adj))
    NPR = max(10, int(200 * sf_adj))
    NCL = max(30, int(3000 * sf_adj))
    NCLL = max(50, int(9000 * sf_adj))
    NDX = max(10, int(100 * sf_adj))


    os.makedirs("data", exist_ok=True)
    con = duckdb.connect("data/warehouse.duckdb")

    con.execute("""
    DROP TABLE IF EXISTS diagnoses; DROP TABLE IF EXISTS claim_lines;
    DROP TABLE IF EXISTS claims; DROP TABLE IF EXISTS providers;
    DROP TABLE IF EXISTS patients;
    CREATE TABLE patients(patient_id INTEGER PRIMARY KEY, dob DATE,
        gender VARCHAR, zip_code VARCHAR, plan_type VARCHAR, state VARCHAR);
    CREATE TABLE providers(provider_id INTEGER PRIMARY KEY, name VARCHAR,
        specialty VARCHAR, state VARCHAR, is_in_network BOOLEAN, npi VARCHAR);
    CREATE TABLE claims(claim_id INTEGER PRIMARY KEY, patient_id INTEGER,
        provider_id INTEGER, service_date DATE, claim_type VARCHAR,
        total_billed DECIMAL(12,2), total_allowed DECIMAL(12,2),
        total_paid DECIMAL(12,2), status VARCHAR, denial_reason VARCHAR);
    CREATE TABLE claim_lines(line_id INTEGER PRIMARY KEY, claim_id INTEGER,
        cpt_code VARCHAR, quantity INTEGER, unit_cost DECIMAL(10,2),
        allowed_amount DECIMAL(10,2), paid_amount DECIMAL(10,2));
    CREATE TABLE diagnoses(diag_id INTEGER PRIMARY KEY, claim_id INTEGER,
        icd_code VARCHAR, is_primary BOOLEAN, chronic_flag BOOLEAN);
    """)

    base = date(2020, 1, 1)
    genders = ["M", "F", "U"]
    plans = ["HMO", "PPO", "EPO", "HDHP", "Medicare", "Medicaid"]
    specialties = ["internal_medicine", "cardiology", "oncology", "orthopedics", "primary_care", "emergency", "radiology", "psychiatry"]
    ctypes = ["professional", "facility", "pharmacy", "dental"]
    cstatuses = ["paid", "denied", "pending", "partial"]
    denial_reasons = ["not_covered", "prior_auth", "out_of_network", None, None, None]
    cpt_codes = [f"CPT{i:05d}" for i in range(1, 51)]
    icd_codes = [f"ICD{i:04d}" for i in range(1, 101)]
    states = ["CA", "TX", "NY", "FL", "IL", "WA", "OH", "GA"]

    cpu_count = get_worker_count()
    progress = GenerationProgress("p07_healthcare", 5)
    with ProcessPoolExecutor(max_workers=cpu_count) as executor:
        progress.advance("patients")
        batched_insert(con, "patients", ['patient_id', 'dob', 'gender', 'zip_code', 'plan_type', 'state'],
                       run_parallel(executor, generate_patients_chunk, NPA, genders, plans, states, base))
        progress.advance("providers")
        batched_insert(con, "providers", ['provider_id', 'name', 'specialty', 'state', 'is_in_network', 'npi'],
                       run_parallel(executor, generate_providers_chunk, NPR, specialties, states))
        progress.advance("claims")
        batched_insert(con, "claims", ['claim_id', 'patient_id', 'provider_id', 'service_date', 'claim_type', 'total_billed', 'total_allowed', 'total_paid', 'status', 'denial_reason'],
                       run_parallel(executor, generate_claims_chunk, NCL, NPA, NPR, base, ctypes, cstatuses, denial_reasons))
        progress.advance("claim_lines")
        batched_insert(con, "claim_lines", ['line_id', 'claim_id', 'cpt_code', 'quantity', 'unit_cost', 'allowed_amount', 'paid_amount'],
                       run_parallel(executor, generate_claim_lines_chunk, NCLL, NCL, cpt_codes))
        progress.advance("diagnoses")
        batched_insert(con, "diagnoses", ['diag_id', 'claim_id', 'icd_code', 'is_primary', 'chronic_flag'],
                       run_parallel(executor, generate_diagnoses_chunk, NDX, NCL, icd_codes))


    con.close()
    print_generation_summary(
        "p07_healthcare",
        sf,
        {
            "patients": NPA,
            "providers": NPR,
            "claims": NCL,
            "claim_lines": NCLL,
            "diagnoses": NDX,
        },
    )

if __name__ == "__main__":
    main()

File: dbgen-rust/src/p07_healthcare.rs
use chrono::{Duration, NaiveDate};
use duckdb::Connection;
use indicatif::{ProgressBar, ProgressStyle};
use rand::prelude::*;
use rand::rngs::SmallRng;

pub fn run(sf: f64, con: &mut Connection) -> duckdb::Result<()> {
    let sf_adj = sf * 4000.0;
    let npa = (1000.0 * sf_adj).max(20.0) as usize;
    let npr = (200.0 * sf_adj).max(10.0) as usize;
    let ncl = (3000.0 * sf_adj).max(30.0) as usize;
    let ncll = (9000.0 * sf_adj).max(50.0) as usize;
    let ndx = (100.0 * sf_adj).max(10.0) as usize;

    con.execute_batch(
        "DROP TABLE IF EXISTS diagnoses; DROP TABLE IF EXISTS claim_lines;
         DROP TABLE IF EXISTS claims; DROP TABLE IF EXISTS providers;
         DROP TABLE IF EXISTS patients;
         CREATE TABLE patients(patient_id INTEGER PRIMARY KEY, dob DATE,
             gender VARCHAR, zip_code VARCHAR, plan_type VARCHAR, state VARCHAR);
         CREATE TABLE providers(provider_id INTEGER PRIMARY KEY, name VARCHAR,
             specialty VARCHAR, state VARCHAR, is_in_network BOOLEAN, npi VARCHAR);
         CREATE TABLE claims(claim_id INTEGER PRIMARY KEY, patient_id INTEGER,
             provider_id INTEGER, service_date DATE, claim_type VARCHAR,
             total_billed DECIMAL(12,2), total_allowed DECIMAL(12,2),
             total_paid DECIMAL(12,2), status VARCHAR, denial_reason VARCHAR);
         CREATE TABLE claim_lines(line_id INTEGER PRIMARY KEY, claim_id INTEGER,
             cpt_code VARCHAR, quantity INTEGER, unit_cost DECIMAL(10,2),
             allowed_amount DECIMAL(10,2), paid_amount DECIMAL(10,2));
         CREATE TABLE diagnoses(diag_id INTEGER PRIMARY KEY, claim_id INTEGER,
             icd_code VARCHAR, is_primary BOOLEAN, chronic_flag BOOLEAN);",
    )?;

    let base_date = NaiveDate::from_ymd_opt(2020, 1, 1).unwrap();
    let genders = ["M", "F", "U"];
    let plans = ["HMO", "PPO", "EPO", "HDHP", "Medicare", "Medicaid"];
    let specialties = [
        "internal_medicine",
        "cardiology",
        "oncology",
        "orthopedics",
        "primary_care",
        "emergency",
        "radiology",
        "psychiatry",
    ];
    let ctypes = ["professional", "facility", "pharmacy", "dental"];
    let cstatuses = ["paid", "denied", "pending", "partial"];
    let denial_reasons = ["not_covered", "prior_auth", "out_of_network"];
    let states = ["CA", "TX", "NY", "FL", "IL", "WA", "OH", "GA"];
    let cpt_codes: Vec<String> = (1..=50).map(|i| format!("CPT{:05}", i)).collect();
    let icd_codes: Vec<String> = (1..=100).map(|i| format!("ICD{:04}", i)).collect();

    let pb = ProgressBar::new(5);
    pb.set_style(
        ProgressStyle::default_bar()
            .template("[{elapsed_precise}] {bar:40.cyan/blue} {pos}/{len} {msg}")
            .unwrap(),
    );

    // 1. Patients
    crate::generate_table_parallel(con, "patients", npa, &pb, "Generating patients...", |i| {
        let mut rng = SmallRng::seed_from_u64(i as u64);
        let dob = base_date - Duration::days(rng.gen_range(365 * 5..365 * 85));
        let gender = genders[rng.gen_range(0..genders.len())];
        let zip = format!("{:05}", rng.gen_range(10000..100000));
        let plan = plans[rng.gen_range(0..plans.len())];
        let state = states[rng.gen_range(0..states.len())];
        (i as i32, dob, gender, zip, plan, state)
    })?;

    // 2. Providers
    crate::generate_table_parallel(con, "providers", npr, &pb, "Generating providers...", |i| {
        let mut rng = SmallRng::seed_from_u64(i as u64);
        let name = format!("Provider {}", i);
        let spec = specialties[rng.gen_range(0..specialties.len())];
        let state = states[rng.gen_range(0..states.len())];
        let network = rng.gen_bool(0.8);
        let npi = format!("NPI{:010}", i);
        (i as i32, name, spec, state, network, npi)
    })?;

    // 3. Claims
    crate::generate_table_parallel(con, "claims", ncl, &pb, "Generating claims...", |i| {
        let mut rng = SmallRng::seed_from_u64(i as u64);
        let pat_id = rng.gen_range(1..=npa) as i32;
        let prov_id = rng.gen_range(1..=npr) as i32;
        let service = base_date + Duration::days(rng.gen_range(0..1096));
        let ctype = ctypes[rng.gen_range(0..ctypes.len())];
        let billed = ((rng.gen_range(100.0..50000.0) * 100.0) as f64).round() / 100.0;
        let allowed = ((billed * rng.gen_range(0.4..0.95) * 100.0) as f64).round() / 100.0;
        let paid = if rng.gen_bool(0.9) {
            ((allowed * rng.gen_range(0.5..1.0) * 100.0) as f64).round() / 100.0
        } else {
            0.0
        };
        let status = cstatuses[rng.gen_range(0..cstatuses.len())];
        let denial = if status == "denied" {
            Some(denial_reasons[rng.gen_range(0..denial_reasons.len())])
        } else {
            None
        };
        (
            i as i32, pat_id, prov_id, service, ctype, billed, allowed, paid, status, denial,
        )
    })?;

    // 4. Claim Lines
    crate::generate_table_parallel(
        con,
        "claim_lines",
        ncll,
        &pb,
        "Generating claim lines...",
        |i| {
            let mut rng = SmallRng::seed_from_u64(i as u64);
            let cl_id = rng.gen_range(1..=ncl) as i32;
            let cpt = &cpt_codes[rng.gen_range(0..cpt_codes.len())];
            let qty = rng.gen_range(1..6);
            let cost = ((rng.gen_range(10.0..5000.0) * 100.0) as f64).round() / 100.0;
            let allowed = ((rng.gen_range(5.0..4000.0) * 100.0) as f64).round() / 100.0;
            let paid = ((rng.gen_range(0.0..3500.0) * 100.0) as f64).round() / 100.0;
            (i as i32, cl_id, cpt.clone(), qty, cost, allowed, paid)
        },
    )?;

    // 5. Diagnoses
    crate::generate_table_parallel(con, "diagnoses", ndx, &pb, "Generating diagnoses...", |i| {
        let mut rng = SmallRng::seed_from_u64(i as u64);
        let cl_id = rng.gen_range(1..=ncl) as i32;
        let icd = &icd_codes[rng.gen_range(0..icd_codes.len())];
        let primary = rng.gen_bool(0.7);
        let chronic = rng.gen_bool(0.4);
        (i as i32, cl_id, icd.clone(), primary, chronic)
    })?;

    pb.finish_with_message("p07_healthcare complete");

    Ok(())
}
